sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -869,26 +873,55 @@ class Analyzer(override val catalogManager: CatalogManager)
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
       _.containsPattern(UNPIVOT), ruleId) {
 
-      // once children and ids are resolved, we can determine values, if non were given
-      case up: Unpivot if up.childrenResolved && up.ids.forall(_.resolved) && up.values.isEmpty =>
-        up.copy(values = up.child.output.diff(up.ids))
+      // once children are resolved, we can determine values from ids and vice versa
+      // if only either is given
+      case up: Unpivot if up.childrenResolved &&
+        up.ids.exists(_.forall(_.resolved)) && up.values.isEmpty =>
+        up.copy(values =
+          Some(
+            up.child.output.diff(up.ids.get.flatMap(_.references))
```

Comment:
Oh I see, it's to match the behavior of when id columns are not specified. It's better to make them consistent, but it's also important to match both Pandas behavior and SQL behavior.

Reply:
I thought, the following would be annoying:
```
scala> val df = spark.range(5).select($"id", ($"id"*10).as("val"))
scala> df.show()
+---+---+
| id|val|
+---+---+
|  0|  0|
|  1| 10|
|  2| 20|
|  3| 30|
|  4| 40|
+---+---+

df.unpivot(Array($"id" * 2), "col", "val").show()
+--------+---+---+
|(id * 2)|col|val|
+--------+---+---+
|       0|id|  0|
|       0|val|  0|
|       2|id| 1|
|       2|val| 10|
|       4|id 2|
|       4|val| 20|
|       6|id 3|
|       6|val| 30|
|       8|id 4|
|       8|val| 40|
+--------+---+---+
```

Of course, that id manipulation can be done before `unpivot` to "materialize" it as a reference:
```
df.withColumn("id", $"id" * 2).unpivot(Array($"id"), "col", "val").show()
+---+---+---+
| id|col|val|
+---+---+---+
|  0|val|  0|
|  2|val| 10|
|  4|val| 20|
|  6|val| 30|
|  8|val| 40|
+---+---+---+
```

Happy to remove that complexity.

Reply:
For DataFrame API, we'd better follow pandas even if it's annoying... SQL API is a different story. Do other databases have the same behavior as this PR does when id columns are not specified?

Reply:
Scala and Python API both require ids. SQL API does not allow to specify ids, which is the same in other SQL databases.

Reply:
I mean, how other databases infer the id columns? It's `allOutput -- valueColumns` or `allOutput -- valueColumns.flatMap(_.references)`?

Reply:
The former.

Reply:
Then why don't we follow other databases? The previous code `up.child.output.diff(up.ids)` is corrected.

Reply:
In contrast to other databases, the Spark implementation allows for expressions rather than column names:

```scala
df.unpivot(
  Array($"id"),
  Array($"struct.val", $"struct.time".cast(DateType).as("date")),
  "column",
  "value"
)
```

```SQL
SELECT *
FROM df
UNPIVOT (
  value FOR column in (struct.val, CAST(struct.time AS DATE) AS date)
)
```

This complicates identifying the unpivoted (value) columns in order to derive the id columns. [Ignoring references here will add columns used in unpivot expressions to the id columns](https://github.com/apache/spark/pull/37407/commits/de19ba0bb33b6369a339a9578173e628102332ae#diff-f306d1dfcd20c4cdcdb5d631416ee9d92bf6893422b0a9683ca4493d7215e5e8):

```SQL
SELECT * FROM courseEarnings
UNPIVOT (
  earningsYear FOR year IN (`2012` as `twenty-twelve`, `2013` as `twenty-thirteen`, `2014` as `twenty-fourteen`)
)

-- !query schema
struct<course:string,2012:int,2013:int,2014:int,year:string,earningsYear:int>

-- !query output
Java	20000	30000	NULL	twenty-fourteen	NULL
Java	20000	30000	NULL	twenty-thirteen	30000
Java	20000	30000	NULL	twenty-twelve	20000
dotNET	15000	48000	22500	twenty-fourteen	22500
dotNET	15000	48000	22500	twenty-thirteen	48000
dotNET	15000	48000	22500	twenty-twelve	15000
```

This is not a problem for the Scala API as we always provide id columns there. But in the SQL API, this gets messy.

Maybe we should disallow expressions for unpivot columns (ids and values). Those expression can be expressed in the FROM clause.

But then the Scala API would also need to abandon expressions, to be like the SQL API. Or would you say they don't need to match in this sense?

Reply:
Alright, I'll create a separate PR for that, and rebase this afterwards.

Reply:
@cloud-fan @HyukjinKwon

Scala `Dataset.groupBy.pivot` allows for expressions, whereas PySpark SQL (`pyspark.sql.DataFrame`) and PySpark Pandas (`pyspark.pandas.DataFrame`) allow for column names only. So we are already mixing expressions and column names.

Scala `Dataset.unpivot` should allow for (named) expressions (as it currently does), to mirror Scala `Dataset.groupBy.pivot`.
SQL should allow for column names only, as all other databases do.
I am fine with downgrading PySpark SQL from allowing `ColumnOrName`to `str` to mirror existing PySpark Pandas `pyspark.pandas.DataFrame.melt`.

Reply:
@EnricoMi Are you working on making this change?

Reply:
Waiting for your reply on https://github.com/apache/spark/pull/37407#discussion_r977818035.

Reply:
Your proposal makes sense to me!

Reply:
Will start with that change now.

Downgrading PySpark SQL to column names will be done in a separate PR, downgrading SQL to column names is going to happen here.

Reply:
This has been done for SQL:  b6b31a0, 4c01645

Reply:
This has been done for PySpark in separate PR: #38020